File: CurrencyFavorites.java
// CurrencyFavorites.java
import java.io.*;
import java.net.*;
import java.nio.file.*;
import java.time.*;
import java.util.*;
import com.google.gson.*;

public class CurrencyFavorites {
    private static final String FAVORITES_FILE = "favorites.json";
    private static final String API_URL = "https://api.frankfurter.app/latest?from=";
    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();

    static class Pair {
        String base;
        String target;
        Pair(String base, String target) { this.base = base; this.target = target; }
    }

    static class Favorites {
        List<Pair> pairs = new ArrayList<>();
    }

    static List<Pair> loadFavorites() throws IOException {
        Path path = Paths.get(FAVORITES_FILE);
        if (Files.exists(path)) {
            String json = new String(Files.readAllBytes(path));
            Favorites fav = gson.fromJson(json, Favorites.class);
            return fav.pairs;
        }
        // default
        return Arrays.asList(new Pair("USD", "EUR"), new Pair("USD", "JPY"));
    }

    static void saveFavorites(List<Pair> pairs) throws IOException {
        Favorites fav = new Favorites();
        fav.pairs = pairs;
        Files.write(Paths.get(FAVORITES_FILE), gson.toJson(fav).getBytes());
    }

    static Map<String, Double> getRates(String base) throws Exception {
        URL url = new URL(API_URL + base);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream()));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) sb.append(line);
        reader.close();
        JsonObject obj = gson.fromJson(sb.toString(), JsonObject.class);
        JsonObject rates = obj.getAsJsonObject("rates");
        Map<String, Double> map = new HashMap<>();
        for (String key : rates.keySet()) {
            map.put(key, rates.get(key).getAsDouble());
        }
        return map;
    }

    static void listRates(List<Pair> pairs) {
        if (pairs.isEmpty()) {
            System.out.println("No favorite pairs.");
            return;
        }
        // group by base
        Map<String, List<Pair>> grouped = new HashMap<>();
        for (Pair p : pairs) {
            grouped.computeIfAbsent(p.base, k -> new ArrayList<>()).add(p);
        }
        System.out.println("\n💱 Favorite Currency Pairs\n");
        for (Map.Entry<String, List<Pair>> entry : grouped.entrySet()) {
            String base = entry.getKey();
            try {
                Map<String, Double> rates = getRates(base);
                for (Pair p : entry.getValue()) {
                    if (rates.containsKey(p.target)) {
                        System.out.printf("%s/%s: %.4f%n", p.base, p.target, rates.get(p.target));
                    } else {
                        System.out.printf("%s/%s: N/A%n", p.base, p.target);
                    }
                }
            } catch (Exception e) {
                System.err.println("Error fetching rates for " + base + ": " + e.getMessage());
            }
        }
        System.out.printf("%nUpdated: %s%n", LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")));
    }

    public static void main(String[] args) throws Exception {
        if (args.length == 0) {
            listRates(loadFavorites());
            return;
        }
        String cmd = args[0].toLowerCase();
        switch (cmd) {
            case "list":
                listRates(loadFavorites());
                break;
            case "add":
                if (args.length != 3) {
                    System.out.println("Usage: add BASE TARGET");
                    return;
                }
                String base = args[1].toUpperCase();
                String target = args[2].toUpperCase();
                List<Pair> pairs = loadFavorites();
                boolean exists = pairs.stream().anyMatch(p -> p.base.equals(base) && p.target.equals(target));
                if (!exists) {
                    pairs.add(new Pair(base, target));
                    saveFavorites(pairs);
                    System.out.printf("✅ Added %s/%s%n", base, target);
                } else {
                    System.out.printf("Pair %s/%s already in favorites.%n", base, target);
                }
                break;
            case "remove":
                if (args.length != 3) {
                    System.out.println("Usage: remove BASE TARGET");
                    return;
                }
                base = args[1].toUpperCase();
                target = args[2].toUpperCase();
                pairs = loadFavorites();
                boolean removed = pairs.removeIf(p -> p.base.equals(base) && p.target.equals(target));
                if (removed) {
                    saveFavorites(pairs);
                    System.out.printf("✅ Removed %s/%s%n", base, target);
                } else {
                    System.out.printf("Pair %s/%s not found.%n", base, target);
                }
                break;
            case "watch":
                int interval = 60;
                for (int i = 1; i < args.length; i++) {
                    if (args[i].equals("--interval") && i+1 < args.length) {
                        interval = Integer.parseInt(args[i+1]);
                        i++;
                    }
                }
                System.out.printf("Watching every %ds. Press Ctrl+C to stop.%n", interval);
                List<Pair> watchPairs = loadFavorites();
                while (true) {
                    listRates(watchPairs);
                    Thread.sleep(interval * 1000L);
                }
            default:
                System.out.println("Usage: java CurrencyFavorites [list|add BASE TARGET|remove BASE TARGET|watch [--interval N]]");
        }
    }
}

File: currency_favorites.go
// currency_favorites.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strconv"
	"strings"
	"time"
)

type Favorites struct {
	Pairs [][2]string `json:"pairs"`
}

var favoritesFile = "favorites.json"
var apiBase = "https://api.frankfurter.app/latest?from="

func loadFavorites() [][2]string {
	data, err := os.ReadFile(favoritesFile)
	if err != nil {
		// default pairs
		return [][2]string{{"USD", "EUR"}, {"USD", "JPY"}}
	}
	var f Favorites
	if err := json.Unmarshal(data, &f); err != nil {
		return [][2]string{{"USD", "EUR"}, {"USD", "JPY"}}
	}
	return f.Pairs
}

func saveFavorites(pairs [][2]string) {
	f := Favorites{Pairs: pairs}
	data, _ := json.MarshalIndent(f, "", "  ")
	os.WriteFile(favoritesFile, data, 0644)
}

func getRates(base string) (map[string]float64, error) {
	url := apiBase + base
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, err
	}
	rates, ok := result["rates"].(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid response")
	}
	ratesMap := make(map[string]float64)
	for k, v := range rates {
		ratesMap[k] = v.(float64)
	}
	return ratesMap, nil
}

func listRates(pairs [][2]string) {
	if len(pairs) == 0 {
		fmt.Println("No favorite pairs.")
		return
	}
	// Group by base
	baseMap := make(map[string][][2]string)
	for _, p := range pairs {
		base := p[0]
		baseMap[base] = append(baseMap[base], p)
	}
	fmt.Println("\n💱 Favorite Currency Pairs\n")
	for base, pairList := range baseMap {
		rates, err := getRates(base)
		if err != nil {
			fmt.Printf("Error fetching rates for %s: %v\n", base, err)
			continue
		}
		for _, p := range pairList {
			target := p[1]
			if rate, ok := rates[target]; ok {
				fmt.Printf("%s/%s: %.4f\n", base, target, rate)
			} else {
				fmt.Printf("%s/%s: N/A\n", base, target)
			}
		}
	}
	fmt.Printf("\nUpdated: %s\n", time.Now().Format("2006-01-02 15:04:05"))
}

func main() {
	if len(os.Args) < 2 {
		pairs := loadFavorites()
		listRates(pairs)
		return
	}
	cmd := os.Args[1]
	switch cmd {
	case "list":
		pairs := loadFavorites()
		listRates(pairs)
	case "add":
		if len(os.Args) != 4 {
			fmt.Println("Usage: add <base> <target>")
			return
		}
		base := strings.ToUpper(os.Args[2])
		target := strings.ToUpper(os.Args[3])
		pairs := loadFavorites()
		newPair := [2]string{base, target}
		exists := false
		for _, p := range pairs {
			if p == newPair {
				exists = true
				break
			}
		}
		if !exists {
			pairs = append(pairs, newPair)
			saveFavorites(pairs)
			fmt.Printf("✅ Added %s/%s\n", base, target)
		} else {
			fmt.Printf("Pair %s/%s already in favorites.\n", base, target)
		}
	case "remove":
		if len(os.Args) != 4 {
			fmt.Println("Usage: remove <base> <target>")
			return
		}
		base := strings.ToUpper(os.Args[2])
		target := strings.ToUpper(os.Args[3])
		pairs := loadFavorites()
		newPairs := [][2]string{}
		found := false
		for _, p := range pairs {
			if p[0] == base && p[1] == target {
				found = true
			} else {
				newPairs = append(newPairs, p)
			}
		}
		if found {
			saveFavorites(newPairs)
			fmt.Printf("✅ Removed %s/%s\n", base, target)
		} else {
			fmt.Printf("Pair %s/%s not found.\n", base, target)
		}
	case "watch":
		interval := 60
		if len(os.Args) >= 4 && os.Args[2] == "--interval" {
			interval, _ = strconv.Atoi(os.Args[3])
		}
		fmt.Printf("Watching every %ds. Press Ctrl+C to stop.\n", interval)
		pairs := loadFavorites()
		ticker := time.NewTicker(time.Duration(interval) * time.Second)
		defer ticker.Stop()
		for {
			listRates(pairs)
			<-ticker.C
		}
	default:
		fmt.Println("Usage: currency_favorites [list|add BASE TARGET|remove BASE TARGET|watch [--interval N]]")
	}
}
